Sources/LLVM/Constant.swift

```diff
@@ -93,17 +71,97 @@ public struct Constant<Repr: ConstantRepresentation>: IRValue {
     let val = self.asLLVM()
-    switch self.repr {
-    case .unsigned:
-      return Constant<Floating>(llvm: LLVMConstUIToFP(val, type.asLLVM()))
-    case .signed:
-      return Constant<Floating>(llvm: LLVMConstSIToFP(val, type.asLLVM()))
-    case .floating:
-      return Constant<Floating>(llvm: LLVMConstFPCast(val, type.asLLVM()))
+    return Constant<Floating>(llvm: LLVMConstUIToFP(val, type.asLLVM()))
+  }
+}
+
+extension Constant where Repr == Signed {
+
+  /// Creates a constant cast to a given integral type.
+  ///
+  /// - parameter type: The type to cast towards.
+  ///
+  /// - returns: A const value representing this value cast to the given
+  ///   integral type.
+  public func cast<T: IntegralConstantRepresentation>(to type: IntType) -> Constant<T> {
+    let destID = ObjectIdentifier(T.self)
+    let val = self.asLLVM()
+    if destID == ObjectIdentifier(Unsigned.self) {
+      return Constant<T>(llvm: LLVMConstIntCast(val, type.asLLVM(), /*signed:*/ false.llvm))
+    } else if destID == ObjectIdentifier(Signed.self) {
+      return Constant<T>(llvm: LLVMConstIntCast(val, type.asLLVM(), /*signed:*/ true.llvm))
+    } else {
+      fatalError("Invalid representation \(type(of: T.self))")
     }
   }
 
-  // MARK: Arithmetic Operations
+  /// Creates a constant cast to a given floating type.
+  ///
+  /// - parameter type: The type to cast towards.
+  ///
+  /// - returns: A const value representing this value cast to the given 
+  ///   floating type.
+  public func cast(to type: FloatType) -> Constant<Floating> {
+    let val = self.asLLVM()
+    return Constant<Floating>(llvm: LLVMConstSIToFP(val, type.asLLVM()))
+  }
+}
+
+extension Constant where Repr == Floating {
+
+  /// Creates a constant cast to a given integral type.
+  ///
+  /// - parameter type: The type to cast towards.
+  ///
+  /// - returns: A const value representing this value cast to the given
+  ///   integral type.
+  public func cast<T: IntegralConstantRepresentation>(to type: IntType) -> Constant<T> {
+    let destID = ObjectIdentifier(T.self)
+    let val = self.asLLVM()
+    if destID == ObjectIdentifier(Unsigned.self) {
+      return Constant<T>(llvm: LLVMConstFPToUI(val, type.asLLVM()))
+    } else if destID == ObjectIdentifier(Signed.self) {
+      return Constant<T>(llvm: LLVMConstFPToSI(val, type.asLLVM()))
+    } else {
+      fatalError("Invalid representation \(type(of: T.self))")
+    }
+  }
+
+  /// Creates a constant cast to a given floating type.
+  ///
+  /// - parameter type: The type to cast towards.
+  ///
+  /// - returns: A const value representing this value cast to the given 
+  ///   floating type.
+  public func cast(to type: FloatType) -> Constant<Floating> {
+    let val = self.asLLVM()
+    return Constant<Floating>(llvm: LLVMConstFPCast(val, type.asLLVM()))
+  }
+}
+
+
+extension Constant where Repr == Struct {
+
+  /// - note: See the LLVM Reference manual's section on `bitcast`
```

Comment:
I would prefer these notes be attached to the unavailability message, but I'm glad these are marked unavailable! Type safety!